sources/host-ctr/pidnamespaces/pid_namespaces.go

```diff
@@ -0,0 +1,169 @@
+	}
+	tmpPidNsList := &HostCtrPidNs{}
+	hostCtrPidNsList := &HostCtrPidNs{}
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &tmpPidNsList)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal PID NS manifest")
+		}
+		for _, ele := range tmpPidNsList.PidNamespaces {
+			// If it already exists in the manifest list for whatever reason, we can just return without doing anything.
+			if ele == pidNs {
+				return nil
+			}
+		}
+	}
+	hostCtrPidNsList.PidNamespaces = append(tmpPidNsList.PidNamespaces, pidNs)
+	if err = writeManifest(hostCtrPidNsList, manifest); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemovePidNsFromManifest takes a PID's namespace and removes it from the internal manifest
+func RemovePidNsFromManifest(pidNs int64) (err error) {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	ok, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil && !ok {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	defer func() {
+		if err = fileLock.Unlock(); err != nil {
+			err = errors.Wrap(err, "failed to release lock for writing to PID NS manifest")
+		}
+	}()
+
+	// Remove task PID NS from manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	hostCtrPidNsList := &HostCtrPidNs{}
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &hostCtrPidNsList)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal PID NS manifest")
+		}
+		for i, ele := range hostCtrPidNsList.PidNamespaces {
+			// Remove PID NS from list
+			if ele == pidNs {
+				hostCtrPidNsList.PidNamespaces = append(hostCtrPidNsList.PidNamespaces[:i], hostCtrPidNsList.PidNamespaces[i+1:]...)
+			}
+		}
+	}
+	if err = writeManifest(hostCtrPidNsList, manifest); err != nil {
+		return err
+	}
+	return nil
+}
+
+// writeManifest writes the list of host-container PID NS to the open manifest file
+func writeManifest(hostCtrPidNsList *HostCtrPidNs, manifest *os.File) error {
+	newData, err := json.Marshal(*hostCtrPidNsList)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal PID NS list to JSON")
+	}
+	if err := manifest.Truncate(0); err != nil {
+		return errors.Wrap(err, "failed to truncate PID NS manifest")
+	}
+	if _, err := manifest.Seek(0, 0); err != nil {
+		return errors.Wrap(err, "failed to reset PID NS manifest I/O offset")
+	}
+	if _, err = manifest.Write(newData); err != nil {
+		return errors.Wrap(err, "failed to write to PID NS manifest")
+	}
```

Comment:
Ideally we'd write a temporary file into the same directory, then copy it over top of the target file.

That way, an application using inotify would see either the old file or the new file. Here I think it might see a zero-byte file or else a partially-written one.